### ISSUE TITLE
Wrong environment variables on a multiuser system

### ISSUE THREAD

**Issue by fabolhak** [ros-industrial/ros_qtc_plugin#209]
Hello,
thank you for the great plugin. It is very useful and helps a lot. Unfortunately we have some problems on our multi-user machine. Somehow the build environment variables gets mixed up and includes paths from other or multiple users (see screenshot). We have our own workspaces in each home directory, but the plugin somehow adds paths even from other users workspaces to the build environment variables.

This introduces a lot of problems for the code model and also the build fails :(.

We are running Ubuntu 16.04 with the version 4.4.1 and qt 5.9 installed.

<img width="814" alt="bildschirmfoto 2017-12-06 um 12 59 52" src="https://user-images.githubusercontent.com/20029691/33660985-3b7f4d26-da86-11e7-8f72-4a129ef10b55.png">

**Comment by Levi-Armstrong** [comment]
That is interesting. Could you try opening a terminal and source the ros and your workspace bash file and then print out the environment and see if you see the same thing?

**Comment by fabolhak** [comment]
Hey, thank you for your fast reply. The ros environment variables seem to be find. After `source /opt/ros/kinetic/setup.bash` and `source /home/fabian/catkin_ws/devel/setup.bash`. I get the following env variables:
```
LC_PAPER=de_DE.UTF-8
MATE_DESKTOP_SESSION_ID=this-is-deprecated
LC_ADDRESS=de_DE.UTF-8
XDG_SESSION_ID=1
LC_MONETARY=de_DE.UTF-8
TERM=xterm
SHELL=/bin/bash
ROS_ROOT=/opt/ros/kinetic/share/ros
SSH_CLIENT=<REMOVED>
ROS_PACKAGE_PATH=/home/fabian/catkin_ws/src:/opt/ros/kinetic/share
X2GO_SESSION=<REMOVED>
QT_LINUX_ACCESSIBILITY_ALWAYS_ON=1
ROS_MASTER_URI=http://localhost:12311
WINDOWID=48235014
LC_NUMERIC=de_DE.UTF-8
GTK_MODULES=gail:atk-bridge:gail:atk-bridge
NX_XINERAMA_CONF=<REMOVED>
X2GO_AGENT_PID=<REMOVED>
USER=fabian
LS_COLORS=<REMOVED>
QT_ACCESSIBILITY=1
LD_LIBRARY_PATH=/home/fabian/catkin_ws/devel/lib:/opt/ros/kinetic/lib:/usr/lib/nx/X11/Xinerama:/usr/lib/nx/X11
LC_TELEPHONE=de_DE.UTF-8
PATH=/opt/ros/kinetic/bin:/home/fabian/bin:/home/fabian/.local/bin:/home/fabian/bin:/home/fabian/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/fabian/.git-radar
MAIL=/var/mail/fabian
LC_IDENTIFICATION=de_DE.UTF-8
PWD=/home/fabian
LANG=en_US.UTF-8
LC_MEASUREMENT=de_DE.UTF-8
QT_GRAPHICSSYSTEM=native
PS1=\[\e[0;32m\]\[\e[m\]\[\e[1;34m\]\w\[\e[m\]$(git-radar --bash --fetch)\[\e[1;34m\]$ \[\e[m\]\[\e[1;37m
XSESSION_EXEC=mate-session
ROSLISP_PACKAGE_DIRECTORIES=/home/fabian/catkin_ws/devel/share/common-lisp
HOME=/home/fabian
SHLVL=4
ROS_DISTRO=kinetic
GTK_OVERLAY_SCROLLING=0
PYTHONPATH=/home/fabian/catkin_ws/devel/lib/python2.7/dist-packages:/opt/ros/kinetic/lib/python2.7/dist-packages
LOGNAME=fabian
LESSOPEN=| /usr/bin/lesspipe %s
PKG_CONFIG_PATH=/home/fabian/catkin_ws/devel/lib/pkgconfig:/opt/ros/kinetic/lib/pkgconfig:/opt/intel/opencl:/opt/intel/opencl
CMAKE_PREFIX_PATH=/home/fabian/catkin_ws/devel:/opt/ros/kinetic
XDG_RUNTIME_DIR=/run/user/1001
DISPLAY=:50
STARTUP=/usr/bin/dbus-launch --exit-with-session /usr/bin/im-launch /usr/bin/env LD_LIBRARY_PATH=/usr/lib/nx/X11/Xinerama:/usr/lib/nx/X11 mate-session
XDG_CURRENT_DESKTOP=MATE
LESSCLOSE=/usr/bin/lesspipe %s %s
LC_TIME=de_DE.UTF-8
ROS_ETC_DIR=/opt/ros/kinetic/etc/ros
COLORTERM=mate-terminal
XAUTHORITY=/home/fabian/.Xauthority
LC_NAME=de_DE.UTF-8
_=/usr/bin/env
```

Our .bashrc has the following entries:
```
source /opt/ros/kinetic/setup.bash
source ~/catkin_ws/devel/setup.bash
export ROS_MASTER_URI=http://localhost:12311
```

**Comment by fabolhak** [comment]
I also noticed: When I manually edit the build environment variables or add a new variable in the QtCreator ROS project, then these changes are gone after reloading the project.

**Comment by Levi-Armstrong** [comment]
Yea, it current does not support user changes. It under the hood sources the workspace for you and update the Qt Environment. 
Can  you tried re-importing the workspace under a different name and see if the environment is the same?

**Comment by fabolhak** [comment]
I did some more tests restarting the system several times: Apparently it matters which user opens the project in QtCreator first after a restart. This workspace path is then being used in the environment variable paths for all other users as well. 

Sometimes we managed to get several workspace paths in the environment vars (see screenshot above). But I didn't find a way to reproduce this yet.

**Comment by Levi-Armstrong** [comment]
Let try creating a new project for the same workspace directory but just give it another name. You will end up with two .workspace and .workspace.user files. Then see if the issue goes away. You may need a separate file for each user in the interim until we can figure out why this is happening.

![image](https://user-images.githubusercontent.com/9803128/33718703-ee336ffe-db23-11e7-8cb9-6fa80413ec58.png)

I think I may understand what is happening based on your description, but I would need you to test a few things. The project that currently has multiple user paths please open its .workspace.user file and find the lines below. In yours the second tag valuelist should have your environment variables, delete them and make sure it looks like what is shown below then then reopen the project and see if the multiple users in the environment are gone and please let me know what happens.

```    <value type="bool" key="ProjectExplorer.BuildConfiguration.ClearSystemEnvironment">false</value>
    <valuelist type="QVariantList" key="ProjectExplorer.BuildConfiguration.UserEnvironmentChanges"/>
    <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Release</value>
```

**Comment by Levi-Armstrong** [comment]
**Note:** there should be multiple places because each build configuration(Debug, Release, etc) has its own environment.

**Comment by fabolhak** [comment]
Hello, thank you for the reply. I'm not sure if I understand your instructions correct, but this is what I did:

### setup
1. creating two usernames `test` and `fabian` on a Ubuntu system 
2. creating different workspaces to test on:
=> personal workspace of `test`: `/home/test/catkin_ws`
=> personal workspace of `fabian`: `/home/fabian/catkin_ws`
=> shared workspace (both users have rw access): `/mnt/data/catkin_ws`
3. reboot

### create project for test
1. login with `test`
2. open QtCreator with ros_qtc_plugin installed and create new ROS project
3. name project `test_project_1` and choose `/home/test/catkin_ws` as path
4. open environment variables section (correct :ballot_box_with_check:):
e.g. `ROS_PACKAGE_PATH=/home/test/catkin_ws/src:/opt/ros/kinetic/share`
5. logout

### create project for fabian
1. login with `fabian`
2. open QtCreator with ros_qtc_plugin installed and create new ROS project
3. name project `test_project_2` and choose `/home/fabian/catkin_ws` as path
4. open environment variables section (not correct ❌):
e.g. `ROS_PACKAGE_PATH=/home/test/catkin_ws/src:/opt/ros/kinetic/share`
5. logout

### create shared project 1
1. login with `test`
2. open QtCreator with ros_qtc_plugin installed and create new ROS project
3. name project `test_project_3` and choose `/mnt/data/catkin_ws` as path
4. open environment variables section:
e.g. `ROS_PACKAGE_PATH=/home/fabian/catkin_ws/src:/opt/ros/kinetic/share` (not correct ❌)
e.g. `CMAKE_PREFIX_PATH=/mnt/data/catkin_ws/devel:/opt/ros/kinetic` (correct :ballot_box_with_check:)
5. logout

6. login with `fabian`
7. open QtCreator with ros_qtc_plugin installed
8. open project `test_project_3`
9. click yes on:
![screenshot_2017-12-08_18-05-29](https://user-images.githubusercontent.com/20029691/33776661-6549a508-dc42-11e7-98c2-2e5abdccf054.png)
10. open environment variables section:
e.g. `ROS_PACKAGE_PATH=/home/fabian/catkin_ws/src:/opt/ros/kinetic/share` (not correct ❌)
e.g. `CMAKE_PREFIX_PATH=/mnt/data/catkin_ws/devel:/opt/ros/kinetic` (correct :ballot_box_with_check:)
11. logout

### create shared project 2
1. login with `fabian`
2. open QtCreator with ros_qtc_plugin installed and create new ROS project
3. name project `test_project_4` and choose `/mnt/data/catkin_ws` as path
4. open environment variables section:
e.g. `ROS_PACKAGE_PATH=/home/fabian/catkin_ws/src:/opt/ros/kinetic/share` (not correct ❌)
e.g. `CMAKE_PREFIX_PATH=/mnt/data/catkin_ws/devel:/opt/ros/kinetic` (correct :ballot_box_with_check:)
5. logout

6. login with `test`
7. analog to previous test

**Comment by fabolhak** [comment]
here are also all the project files: [project_files.zip](https://github.com/ros-industrial/ros_qtc_plugin/files/1543374/project_files.zip)

**Comment by fabolhak** [comment]
> In yours the second tag valuelist should have your environment variables, delete them and make sure it looks like what is shown below then then reopen the project and see if the multiple users in the environment are gone and please let me know what happens.

If I modify anything in these files, then my changes will be overridden the next time I open the project in QtCreator 🙁.

**Comment by fabolhak** [comment]
I also noticed that it is very different what environment variables are included in the ROS project.

For the `test` user the following vars are in the build environment:
![screenshot_2017-12-08_18-23-16](https://user-images.githubusercontent.com/20029691/33777468-3c937186-dc45-11e7-80fd-ed78870b8493.png)

For the `fabian` user the following vars are in the build environment:
![screenshot_2017-12-08_18-23-52](https://user-images.githubusercontent.com/20029691/33777477-44d2ed2c-dc45-11e7-96b2-5cd72f0ce1bc.png)

This is very confusing.

**Comment by Levi-Armstrong** [comment]
Thanks for the test cases. I will look into it and see what I can figure out.

**Comment by Levi-Armstrong** [comment]
@fabolhak This is my remaining issues before the next release so I am going to address this over the next few days.

**Comment by fabolhak** [comment]
If you need anything. Please let me know :).

**Comment by Levi-Armstrong** [comment]
This is addressed by PR #234.

**Comment by Levi-Armstrong** [comment]
@fabolhak I have just released the latest version. Please see wiki for new install procedure.
